--- a/amd/comgr/test/CMakeLists.txt
+++ b/amd/comgr/test/CMakeLists.txt
@@ -59,7 +59,7 @@
 macro(add_test_input_bitcode_bundle name input output)
   add_custom_command(
     OUTPUT "${output}"
-    COMMAND "$<TARGET_FILE:clang>" -c --offload-arch=gfx900 -emit-llvm -fgpu-rdc
+    COMMAND "$<TARGET_FILE:clang>" -c --rocm-device-lib-path=${CMAKE_BINARY_DIR}/../device-libs/amdgcn/bitcode --offload-arch=gfx900 -emit-llvm -fgpu-rdc
     --gpu-bundle-output ${ARGN} "${CMAKE_CURRENT_SOURCE_DIR}/${input}"
     -o "${output}"
     VERBATIM
@@ -76,7 +76,7 @@
 macro(add_test_input_object_file_bundle name input output)
   add_custom_command(
     OUTPUT "${output}"
-    COMMAND "$<TARGET_FILE:clang>" -c --offload-arch=gfx900 --gpu-bundle-output
+    COMMAND "$<TARGET_FILE:clang>" -c --rocm-device-lib-path=${CMAKE_BINARY_DIR}/../device-libs/amdgcn/bitcode --offload-arch=gfx900 --gpu-bundle-output
     --offload-device-only ${ARGN} "${CMAKE_CURRENT_SOURCE_DIR}/${input}"
     -o "${output}"
     VERBATIM
