diff --git a/library/src/amd_detail/hipsparse.cpp b/library/src/amd_detail/hipsparse.cpp
index b044a71b..0b1ce744 100644
--- a/library/src/amd_detail/hipsparse.cpp
+++ b/library/src/amd_detail/hipsparse.cpp
@@ -22,6 +22,9 @@
 * ************************************************************************ */
 
 #include "hipsparse.h"
+#ifndef __glibcxx_assert
+#define __glibcxx_assert(_Condition)
+#endif
 
 #include <hip/hip_complex.h>
 #include <hip/hip_runtime_api.h>
